src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -47,6 +47,8 @@ public class LockFileBuilder
                 Version = _lockFileVersion
             };
 
+            var previousLibraries = previousLockFile?.Libraries.ToDictionary(l => Tuple.Create(l.Name, l.Version));
```

Comment:
Move it inside `else if (library.Type == LibraryType.Package)` (line 109)

Reply:
The type check is done inside a loop on target graphs, and packages. Building the library set is done from the libraries section that is global to the file. All target graphs use the same index.